0010-zkvm-keccak.md

```diff
@@ -0,0 +1,788 @@
+
+## Detailed design
+
+The Keccak hash function (later standardized by NIST as SHA3, with small variations) is known to be a quantum-safe, and very efficient hash to compute. Nonethelss, its intensive use of boolean operations which are costly in finite field arithmetic makes it a SNARK-unfriendly function. In this section, we propose an optimized design for our proof system that should outperform our current Keccak to become usable in the zkVM for the OP stack.
+
+The [Appendix](#appendix) covers a practical description of the Keccak algorithm and specifies the required parameters needed for compatibility in our use case. The configuration of Keccak must be the same as that of the Ethereum EVM, meaning specifically:
+
+> * [pre-NIST variant of Keccak](https://keccak.team/files/Keccak-reference-3.0.pdf) (uses the $10^*1$ padding rule)
+> * variable input **byte** length (not any bit length)
+> * output length of 256 bits
+> * input and output in big endian
+> * state composed of 25 (5x5) 64-bit words
+> * making a block length of 1600 bits ($b$)
+> * 512 bits capacity ($c$)
+> * 1088 length bit rate ($r := b - c$)
+> * 24 rounds of permutation
+
+This section focuses on the actual changes for the proposed gadget.
+
+### Input-Output format
+
+The Keccak gadget is triggered from the sys call. Inputs need to be written into memory and outputs are read from the sys call. One should expect inputs and outputs to be found in bytes serialization, with a big-endian encoding. One can only read from Canon 4 bytes at a time, what means that the hash function will have to wait until the input is fully loaded onto memory.
+
+### Bitwise-sparse representation
+
+Here, we present the bitwise-sparse representation of bitstrings that will be used to obtain more efficient computations of boolean functions within our proof system. 
+
+Given that simple binary operations are very costly within a SNARK, the new idea consists on arithmetizing boolean operations as much as possible, becoming more efficient with finite field algebra. We introduce the _bitwise-sparse representation_ of bitstrings for this goal.
+
+#### Expansion
+
+With this notation, input bitstrings are expanded inserting three `0` bits in between "real" bits (hence the _sparsity_). The motivation for this, is to use intermediate bits to store auxiliary information such as carry bits, resulting from algebraic computations of boolean functions. That means, for each bit in the original bitstring, its expansion will contain four bits. The following expression represents such mapping:
+
+$$\forall X \in (b_i)^\ell, b\in[0,1]: expand(X) \to 0 \ 0 \ 0 \ b_{\ell-1} \ . \ . \ . \ 0 \ 0 \ 0 \ b_0$$
```

Comment:
Opiniated comment: I liked the notation used in zcash to mention the size of the bitstring. They use $$\mathbb{B}^{64} \rightarrow \mathbb{B}^{256}$$ for a function mapping a 64 bit string to a 256 bit string. It helps to categorize the input/outputs by just reading the domain and image of the function.